back/app/controllers/web_api/v1/projects_controller.rb

```diff
@@ -52,6 +52,73 @@ def index
     )
   end
 
+  # For use with 'Open to participation' homepage widget.
+  # Returns all published or archived projects that are visible to user
+  # and in an active participatory phase (where user can do something).
+  # Ordered by the end date of the current phase, soonest first (nulls last).
+  def index_projects_with_active_participatory_phase
+    # Projects user can see, with active participatory (not information) phase & include the phases.end_at column
+    # We could use the ProjectPermissionSevice step to filter for active phases, but doing it here is more efficient.
+    subquery = policy_scope(Project)
+      .joins('INNER JOIN admin_publications AS admin_publications ON admin_publications.publication_id = projects.id')
+      .where(admin_publications: { publication_status: 'published' })
+      .joins('INNER JOIN phases AS phases ON phases.project_id = projects.id')
+      .where(
+        'phases.start_at <= ? AND (phases.end_at >= ? OR phases.end_at IS NULL) AND phases.participation_method != ?',
+        Time.zone.now.to_fs(:db), Time.zone.now.to_fs(:db), 'information'
+      )
+      .select('projects.*, phases.end_at AS phase_end_at')
+
+    # Perform the SELECT DISTINCT on the outer query
+    @projects = Project
+      .from(subquery, :projects)
+      .distinct
+      .order('phase_end_at ASC NULLS LAST')
+      .preload(phases: { permissions: [:groups] })
+
+    # Projects user can participate in, or where such participation could (probably) be made possible by user
+    # (e.g. user not signed in).
+    # Unfortunately, this breaks the query chain, so we have to start a new one after this.
+    # Also, we will need the action descriptors again later, so we will store them.
+
+    # # Step 1: Create pairs of project ids and action descriptors.
+    # # Since this is the last filtering step, we will keep going
+    # # until we reach the limit required for pagination.
+    pagination_limit = calculate_pagination_limit
+
```

Comment:
@jamesspeake I wrote this solution to avoid calculating all action descriptors for all projects and then filtering ones where participation might be possible. This way, we only calculate as many action descriptors as needed until we reach the pagination limit.

There are still other performance improvements to be made. It could be great if we could know in advance for which projects we don't have to check the action descriptors at all. For example, if the current phase does not have any permissions including smart groups, and since we already check visibility and authentication / verification is seen as a "fixable" requirement, maybe we could make the assumption that for such projects we don't need to check the action descriptors at all? What do you think?

This would allow us to make two further improvements:

1. We could avoid loading all projects into memory by just loading the minimum number that will allow us to fill the pagination. E.g. if `page[size]` is `6`, and `page[number]` is `1`. and the first 6 projects in the query all have no permissions with smart groups, we can just take the first 6 and not calculate any action descriptors at all. We could even not load anything into memory in this case. Or, if e.g. the first 5 projects have no smart groups, but project 6 and 7 do, but 8 doesn't, we can just load the first 8 projects into memory, because even if projects 6 and 7 end up filtered out we will still always reach the threshold of 6.
2. And we would only have to calculate the action descriptors for the ones without smart groups of course.

Not sure if my assumption makes sense, and if this type of operation is possible in SQL- just thinking out loud

Reply:
I think this could get even more complex if we were to do that and I think the performance is probably OK as it is.

Reply:
Is this expected to be visible when you're not logged in? If so then I think it might be wise to add a cache call around the whole thing if there is no logged in user

Reply:
> Is this expected to be visible when you're not logged in? If so then I think it might be wise to add a cache call around the whole thing if there is no logged in user

@jamesspeake You mean a BE cache or FE? (Yes, will be visible to non-logged-in visitor, if would show any project(s))